File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bfs/Problem16234.java
package com.keunwon.algorithm.bfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Problem16234 {

    static int N;
    static int L, R;
    static int[][] map;

    static Set<Position> unions = new HashSet<>();
    static int[][] moves = { {-1, 0}, {1, 0}, {0, -1}, {0, 1} };


    public static void main(String[] args) throws IOException {
        input();

        int days = 0;
        while (true) {
            boolean[][] visited = new boolean[N][N];
            boolean isMove = false;

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (visited[i][j]) {
                        continue;
                    }

                    if (bfs(i, j, visited)) {
                        isMove = true;
                    }
                }
            }

            if (isMove) { days++; }
            else { break; }
        }

        System.out.println(days);
    }

    static boolean bfs(int x, int y, boolean[][] visited) {
        Queue<Position> queue = new LinkedList<>();
        List<Position> unions = new ArrayList<>();

        queue.add(new Position(x, y));
        unions.add(new Position(x, y));
        visited[x][y] = true;

        int sum = map[x][y];
        while (!queue.isEmpty()) {
            Position p = queue.poll();

            for (int[] move : moves) {
                int nx = p.x + move[0];
                int ny = p.y + move[1];

                if (!validMap(nx, ny) || visited[nx][ny]) { continue; }

                if (!isGo(map[p.x][p.y], map[nx][ny])) { continue; }

                queue.add(new Position(nx, ny));
                unions.add(new Position(nx, ny));
                sum += map[nx][ny];
                visited[nx][ny] = true;
            }
        }

        if (unions.size() == 1) { return false; }

        int temp = sum / unions.size();
        for (Position union : unions) {
            map[union.x][union.y] = temp;
        }

        return true;
    }

    static boolean isGo(int num, int nextNum) {
        int temp = Math.abs(nextNum - num);
        return L <= temp && temp <= R;
    }

    static boolean validMap(int x, int y) {
        return x >= 0 && x < N && y >= 0 && y < N;
    }

    static void input() throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        L = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        map = new int[N][N];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < N; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }
    }

    static class Position {
        int x;
        int y;

        public Position(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem16234.kt
package com.keunwon.algorithm.baekjoon

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.LinkedList
import kotlin.math.abs

/**
 * <p>
 * 이름: 인구이동
 * 난이도: 골드-4
 * </p>
 *
 * @see <a href="https://www.acmicpc.net/problem/16234">인구이동 (골드-4)</a>
 **/
class Problem16234 {
    fun solution(l: Int, r: Int, map: Array<IntArray>): Int {
        var result = 0
        val range = l..r

        while (true) {
            val visited = Array(map.size) { BooleanArray(map[0].size) }
            var flag = false

            for (i in map.indices) {
                for (j in map[i].indices) {
                    if (!visited[i][j] && bfs(range, Pair(i, j), map, visited)) {
                        flag = true
                    }
                }
            }

            if (!flag) break
            ++result
        }
        return result
    }

    private fun bfs(
        range: IntRange,
        p: Pair<Int, Int>,
        map: Array<IntArray>,
        visited: Array<BooleanArray>,
    ): Boolean {
        val q = LinkedList<Pair<Int, Int>>().apply { offer(p) }
        val nodes = mutableListOf<Pair<Int, Int>>()

        visited[p.first][p.second] = true

        while (q.isNotEmpty()) {
            val (r, c) = q.poll()
            nodes.add(Pair(r, c))

            for ((mr, mc) in moves) {
                val nr = r + mr
                val nc = c + mc

                if (nr !in map.indices || nc !in map[0].indices) continue
                if (visited[nr][nc] || abs(map[r][c] - map[nr][nc]) !in range) continue

                visited[nr][nc] = true
                q.offer(nr to nc)
            }
        }

        if (nodes.size == 1) return false

        val num = nodes.sumOf { (r, c) -> map[r][c] } / nodes.size
        nodes.forEach { (r, c) -> map[r][c] = num }
        return true
    }

    companion object {
        private val moves = arrayOf(-1 to 0, 0 to 1, 1 to 0, 0 to -1)
    }
}

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (n, l, r) = readLine().split(" ").map { it.toInt() }
    val map = Array(n) {
        val arr = readLine().split(" ")
        IntArray(n) { arr[it].toInt() }
    }

    println(Problem16234().solution(l, r, map))
}
